content/en/docs/advanced-tasks/fail-forward-upgrades.md

```diff
@@ -0,0 +1,119 @@
+
+### When one or more CSVs have failed
+
+Let's review what steps must be taken to recover from a failed CSV today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The CSV for `Operator v2` enters the FAILED phase.
+3. `Operator v3` is added to the catalog which replaces or skips `Operator v2`.
+3. The resolver cannot upgrade `Operator v2` while including `Operator v1` in the solution set, upgrade is blocked.
+4. User manually deletes the existing CSVs, a new InstallPlan is generated and approved.
+5. `Operator v3` is installed and the upgrade succeeds.
+
+To further understand, let's visualize this.
+
+{{<mermaid>}}
+graph TD;
+  A(Operator V1 installed) --> B(Starts upgrading to Operator V2)
+
+  B --> |Install succeeds| F(Operator V2 is installed and succeeds)
+
+  B --> |CSV enters FAILED phase| C(Operator V3 added to catalog replacing V2)
+  C --> |Upgrade blocked as V1 is in solution set still| D(Manually delete Operator V2 CSV)
+  D --> |New InstallPlan generated and approved| E(Operator V3 is installed and succeeds)
+
+  B --> |CSV enters FAILED phase with fail forward upgrades| H(Operator V3 added to catalog replacing V2)
+  H --> |New InstallPlan generated and approved| J(Operator V3 is installed and succeeds)
+{{</mermaid>}}
+
+When you opt into "fail forward" upgrades, OLM:
+- Does not include CSVs in the REPLACING phase in the set of arguments sent to the Resolver.
+- Allows CSVs to move from the FAILED phase to the REPLACING phase.
+
+These changes allow OLM to recover from any number of failed CSVs in an upgrade path. You only need to update the CatalogSource with new operators that replace the FAILED CSVs. 
+
+## Opting into Fail Forward Upgrades
+
+Since resolution is namespaced, the toggle for allowing "fail forward" upgrades is namespaced as well to avoid having extra burden on the resolver. This removes complexity surrounding the resolver's understanding of which dependent operators can or cannot use the forced upgrade toggle.
+
+With this in mind, the namespaced resource, OperatorGroup, has the `upgradeStrategy` field.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+exdx marked this conversation as resolved.
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  upgradeStrategy:
+     # Possible values include "Default" or "UnsafeFailForward".
+    name: UnsafeFailForward
+```
+
+With the upgradeStrategy type set to `UnsafeFailForward`, OLM will allow operators to "fail forward" in adherence with the principles discussed below. If the upgradeStrategy is unset or set to `Default` OLM will exhibit existing behavior. 
+
+### __**Warning**__
```

Comment:
Any idea on whether we should be documenting something like this much further up in the file?

Reply:
Yeah, definitely. Raised it to the top.